Use timeout instead of interval for tile prefetching.

gokaysatir opened the pull request:
Issue: interval calls itself multiple times.


Change-Id: I5ef46145debb02c2dd0d716b0adb0dc715449885


* Resolves: # <!-- related github issue -->
* Target version: master 

### Summary


### TODO

- [ ] ...

### Checklist

- [ ] I have run `make prettier-write` and formatted the code.
- [ ] All commits have Change-Id
- [ ] I have run tests with `make check`
- [ ] I have issued `make run` and manually verified that everything looks okay
- [ ] Documentation (manuals or wiki) has been updated or is not required

caolanm commented:
I believe this started to be a problem since commit 66f2449f3e4f00031e4bea86e1a28921a65e9de5 so we probably should revisit what changed there as there wasn't intended to be any change in behavior IIUC

gokaysatir commented:
> I believe this started to be a problem since commit [66f2449](https://github.com/CollaboraOnline/online/commit/66f2449f3e4f00031e4bea86e1a28921a65e9de5) so we probably should revisit what changed there as there wasn't intended to be any change in behavior IIUC

I only removed:
`this._previewInvalidations.push(invalidBounds);`

from Writer and Calc tile layers. Because they don't have previews. That code was not needed in those places. That may have exposed the issue.

mmeeks commented:
Oh! I shortened the timeouts for pre-fetching left and right. Hmm - is it the case that we can multiple pre-fetch the same tiles if they have not been received yet ? if so - "that would be bad"(TM). Gokay- did you test this with the preload map enabled - to ensure that we continue to pre-load properly ? =) - particularly in calc.

gokaysatir commented:
> Oh! I shortened the timeouts for pre-fetching left and right. Hmm - is it the case that we can multiple pre-fetch the same tiles if they have not been received yet ? if so - "that would be bad"(TM). Gokay- did you test this with the preload map enabled - to ensure that we continue to pre-load properly ? =) - particularly in calc.

Yes i tested it now. I think this is a bug that we didn't notice before.

caolanm commented:
I added some comments in https://github.com/CollaboraOnline/online/pull/11211 I think I see some discrepancies in the conversion there that look worth clarifying.

gokaysatir commented:
> I added some comments in #11211 I think I see some discrepancies in the conversion there that look worth clarifying.

Thanks again: You are right, there were 2 issues, fixed now:
* https://github.com/CollaboraOnline/online/pull/11231

I think that we can replace interval with timer, since we already reset prefetcing via various events like scroll and move. I think a position change (move) can be enough for being sure that we prefetch at the right time.

I think this is correct that we re-send the tile request until they arrive:

> Hmm - is it the case that we can multiple pre-fetch the same tiles if they have not been received yet ? 

And it can slow down a bit more when a large document is viewed.

mmeeks commented:
> I think this is correct that we re-send the tile requests until they arrive:

I suspect that is a bug - particularly if we don't have Szymon's patch that stops us re-rendering the tiles as they are requested each time from a tilecombine =) ie. we badly need:

commit 4c85039d7d85424c98c9aeb8e6c3809beed17d4d
Author: Szymon Kłos <szymon.klos@collabora.com>
Date:   Mon Feb 10 16:08:49 2025 +0100

    tiles: do not duplicate rendering if tilecombine requested

in addition there I think.

gokaysatir commented:
I think we will get this in, can we approve the request?

gokaysatir commented:
> Please be careful. This breaks pre-loading.
> 
> ![image](https://private-user-images.githubusercontent.com/122848/417249912-78ea3b23-aa86-4619-b0ae-4e3e79d6565f.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDA2Mzk3NjQsIm5iZiI6MTc0MDYzOTQ2NCwicGF0aCI6Ii8xMjI4NDgvNDE3MjQ5OTEyLTc4ZWEzYjIzLWFhODYtNDYxOS1iMGFlLTRlM2U3OWQ2NTY1Zi5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwMjI3JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDIyN1QwNjU3NDRaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1jN2VmOGM5ODRmNDY1NWQzMjVhZTM5MTEzZTgxYjcxZTdmNmJiYmU3YzA4YmJmN2E1YTk4NjU2NmE4YWVlZWQ1JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.nzMSdnoBRovFBhq6smn20dklAVtmmiYH1Th2Yn7r7IM)
> 
> Please turn on the feature I asked for - and see what happens to incremental tile pre-loading around the display area - which is an iterative process.

I see, i think i mis-read the expected behaviour and couldn't see the regression. Let me abandon this PR. I will cherry-pick the cypress fix.
